spiceypy/spiceypy.py

```diff
@@ -1742,18 +1774,18 @@ def dafgsr(handle, recno, begin, end):
     :type begin: int
     :param end: Last word to read from record.
     :type end: int
-    :return: Contents of record.
-    :rtype: float
+    :return: Contents of request sub-record
+    :rtype: float numpy.ndarray
     """
     handle = ctypes.c_int(handle)
     recno = ctypes.c_int(recno)
     begin = ctypes.c_int(begin)
     end = ctypes.c_int(end)
-    data = ctypes.c_double()
+    data = stypes.emptyDoubleVector(125)
     found = ctypes.c_bool()
-    libspice.dafgsr_c(handle, recno, begin, end, ctypes.byref(data),
+    libspice.dafgsr_c(handle, recno, begin, end, data,
                       ctypes.byref(found))
-    return data.value, found.value
+    return stypes.cVectorToPython(data)[:end.value+1-begin.value], found.value
```

Comment:
The summary record, and all DAF records, are 128 words, i.e. fWords(1:128) in FORTRAN, and pWords[0:128] in Python.

Assuming for the moment that begin and end are in the range 1 to 128 inclusive, and dropping the .value suffix for readability:

This call loads the [begin.value-1:end.value] (Python notation) window from the record starting at the address (data+0) (C notation - passed to dafgsr_c, it  does *not* load data starting at (data+begin-1) (C notation).  So in all cases, data[begin-1:whatever] (Python notation) is wrong.

I picked 125 because the summary record uses 3 words at the beginning for the linked list and single summary count, I realized I should have made it 128 to ensure the vector was big enough to avoid a segfault in all cases.

The better thing to do is to initially size the data vector appropriately i.e.

    data = stypes.emptyDoubleVector(1 + min(1+[end,128]) - max([begin,1]))

Then there is no reason to truncate the data on return.  I will change this wrapper to do that.

Although an argument could be made that whatever the user passes implies that they expect a vector of lenght (1+end-begin); if I was doing this in C, I would provide data as a pointer to that much memory available..

If begin >= end, then I suspect stypes.emptyDoubleVector, or something it calls, will throw an exception, which is fine as it is a user error, so no need to check that here.

Reply:
Note that my original approach is consistent with how you coded dafgs; maybe that is why I did it that way.